pkg/rules/manager.go

```diff
@@ -0,0 +1,349 @@
+					Health:                    string(rule.Health()),
+					LastError:                 lastError,
+					EvaluationDurationSeconds: rule.GetEvaluationDuration().Seconds(),
+					LastEvaluation:            rule.GetEvaluationTimestamp(),
+				}}})
+		default:
+			// We cannot do much, let's panic, API will recover.
+			panic(fmt.Sprintf("rule %q: unsupported type %T", r.Name(), rule))
+		}
+	}
+	return ret
+}
+
+func activeAlertsToProto(s storepb.PartialResponseStrategy, a *rules.AlertingRule) []*rulespb.AlertInstance {
+	active := a.ActiveAlerts()
+	ret := make([]*rulespb.AlertInstance, len(active))
+	for i, ruleAlert := range active {
+		ret[i] = &rulespb.AlertInstance{
+			PartialResponseStrategy: s,
+			Labels:                  &rulespb.PromLabels{Labels: storepb.PromLabelsToLabels(ruleAlert.Labels)},
+			Annotations:             &rulespb.PromLabels{Labels: storepb.PromLabelsToLabels(ruleAlert.Annotations)},
+			State:                   rulespb.AlertState(ruleAlert.State),
+			ActiveAt:                &ruleAlert.ActiveAt,
+			Value:                   strconv.FormatFloat(ruleAlert.Value, 'e', -1, 64),
+		}
+	}
+	return ret
+}
+
+// configRuleGroups is what is parsed from config.
+type configRuleGroups struct {
+	Groups []configRuleGroup `yaml:"groups"`
+}
+
+type configRuleGroup struct {
+	rulefmt.RuleGroup
+	PartialResponseStrategy *storepb.PartialResponseStrategy
+}
+
+// Manager is a partial response strategy and proto compatible Manager.
+// Manager also implements rulespb.Rules gRPC service.
+type Manager struct {
+	workDir string
+	mgrs    map[storepb.PartialResponseStrategy]*rules.Manager
+
+	mtx       sync.RWMutex
+	ruleFiles map[string]string
+}
+
+// NewManager creates new Manager.
+// QueryFunc from baseOpts will be rewritten.
+func NewManager(ctx context.Context, reg prometheus.Registerer, dataDir string, baseOpts rules.ManagerOptions, queryFuncCreator func(partialResponseStrategy storepb.PartialResponseStrategy) rules.QueryFunc) *Manager {
+	m := &Manager{
+		workDir:   filepath.Join(dataDir, tmpRuleDir),
+		mgrs:      make(map[storepb.PartialResponseStrategy]*rules.Manager),
+		ruleFiles: make(map[string]string),
+	}
+	for _, strategy := range storepb.PartialResponseStrategy_value {
+		s := storepb.PartialResponseStrategy(strategy)
+
+		opts := baseOpts
+		opts.Registerer = extprom.WrapRegistererWith(prometheus.Labels{"strategy": strings.ToLower(s.String())}, reg)
+		opts.Context = ctx
+		opts.QueryFunc = queryFuncCreator(s)
+
+		m.mgrs[s] = rules.NewManager(&opts)
+	}
+
+	return m
+}
+
+func (m *Manager) Run() {
+	for _, mgr := range m.mgrs {
+		mgr.Run()
+	}
+}
+
+func (m *Manager) Stop() {
+	for _, mgr := range m.mgrs {
+		mgr.Stop()
+	}
+}
+
+func (m *Manager) protoRuleGroups() []*rulespb.RuleGroup {
+	var res []*rulespb.RuleGroup
+	for _, g := range m.RuleGroups() {
+		res = append(res, g.toProto())
+	}
+	return res
+}
+
+func (m *Manager) RuleGroups() []Group {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+	var res []Group
```

Comment:
consider preallocating slice

Reply:
I cannot, I don't know number that managers have.